pkg/controller/test-images-distributor/test_images_distributor_test.go

```diff
@@ -784,3 +786,81 @@ type noOpRegistryResolver struct{}
 func (noOpRegistryResolver) ResolveConfig(cfg api.ReleaseBuildConfiguration) (api.ReleaseBuildConfiguration, error) {
 	return cfg, nil
 }
+
+func TestSourceForConfigChangeChannel(t *testing.T) {
+	t.Parallel()
+
+	type requestWithCluster struct {
+		cluster string
+		request types.NamespacedName
+	}
+
+	buildClusters := sets.NewString("build01", "build02")
+	testCases := []struct {
+		name   string
+		change agents.IndexDelta
+
+		expected []requestWithCluster
+	}{
+		{
+			name:   "Config was added, we trigger an event per cluster",
+			change: agents.IndexDelta{IndexKey: "namespace/name:tag", Added: []*api.ReleaseBuildConfiguration{{}}},
+
+			expected: []requestWithCluster{
+				{cluster: "build01", request: types.NamespacedName{Namespace: "namespace", Name: "name:tag"}},
+				{cluster: "build02", request: types.NamespacedName{Namespace: "namespace", Name: "name:tag"}},
+			},
+		},
+		{
+			name:   "Config was removed, we don't trigger an event",
+			change: agents.IndexDelta{IndexKey: "namespace/name:tag", Removed: []*api.ReleaseBuildConfiguration{{}}},
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			changeChannel := make(chan agents.IndexDelta)
+
+			source := sourceForConfigChangeChannel(buildClusters, changeChannel)
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			queue := &hijackingQueue{}
+
+			if err := source.InjectStopChannel(ctx.Done()); err != nil {
+				t.Fatalf("failed to inject stop channel into source: %v", err)
+			}
+			if err := source.Start(ctx, &handler.EnqueueRequestForObject{}, queue); err != nil {
+				t.Fatalf("failed to start source: %v", err)
+			}
+			changeChannel <- tc.change
+			close(changeChannel)
+
+			// The source copies the data from the input channel to the workqueue asynchronously
+			// and there is no way for us to properly synchronize that so instead we just wait.
+			// We can not poll because for negative testcases (expect no request) that would instanly
+			// succeed and later events won't make the test fail correctly.
+			time.Sleep(10 * time.Second)
```

Comment:
I guess technically you could have `sourceForConfigChangeChannel` send the events to both the source and to another optional given channel, which would be set only in test code, to allow synchronization. Since this is all in parallel it should not matter to overall unit test runtime though and the current impl is good